src/config.ts

```diff
@@ -0,0 +1,58 @@
+import * as path from 'path';
+import * as process from 'process';
+import * as vscode from 'vscode';
+
+// Gets the config value `clangd.<key>`. Applies ${variable} substitutions.
+export function get<T>(key: string): T {
+  return substitute(vscode.workspace.getConfiguration('clangd').get(key));
+}
+
+// Sets the config value `clangd.<key>`. Does not apply substitutions.
+export function update<T>(key: string, value: T,
+                          target?: vscode.ConfigurationTarget) {
+  return vscode.workspace.getConfiguration('clangd').update(key, value, target);
+}
+
+// Traverse a JSON value, replacing placeholders in all strings.
+function substitute<T>(val: T): T {
+  if (typeof val == 'string') {
+    val = val.replace(/\$\{(.*?)\}/g, (match, name) => {
+      const rep = replacement(name);
+      // If there's no replacement,available, keep the placeholder.
+      return (rep === null) ? match : rep;
+    }) as unknown as T;
+  } else if (Array.isArray(val))
+    val = val.map((x) => substitute(x)) as unknown as T;
+  else if (typeof val == 'object') {
+    // Substitute values but not keys, so we don't deal with collisions.
+    const result = {} as {[k: string]: any};
+    for (let [k, v] of Object.entries(val))
+      result[k] = substitute(v);
+    val = result as T;
+  }
+  return val;
+}
+
+// Subset of substitution variables that are most likely to be useful.
+// https://code.visualstudio.com/docs/editor/variables-reference
+function replacement(name: string): string|null {
+  if (name == 'workspaceRoot' || name == 'workspaceFolder') {
+    if (vscode.workspace.rootPath !== undefined)
+      return vscode.workspace.rootPath;
+    if (vscode.window.activeTextEditor !== undefined)
+      return path.dirname(vscode.window.activeTextEditor.document.uri.fsPath);
+    return process.cwd();
+  }
+  if (name == 'cwd')
+    return process.cwd();
```

Comment:
so if I open `vscode-clangd` project in vscode, `${cwd}` should be `/path/to/vscode-clangd`, but `process.cwd()` returns a `/` on my local machine :(

Reply:
I can't reproduce this. Are you opening `vscode` from the terminal? If you're opening it from a GUI (e.g. Code.app in Finder) I wouldn't expect the working directory to be meaningful

Reply:
Indeed, thanks for pointing this out.
I've made it do the same as workspaceRoot which seems to be basically right except for the error cases.